coalals/concurrency.py

```diff
@@ -0,0 +1,110 @@
+from .utils.wrappers import func_wrapper
+from concurrent.futures import ProcessPoolExecutor
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+class JobTracker:
+    """
+    JobTracker helps to keep track of running jobs and
+    function as an advanced counter. It is lazy and will
+    only updated its internal state when requested for
+    information about it. JobTracker instances should always
+    live on the main thread. Hence should never go out of sync.
+    """
+
+    @staticmethod
+    def kill_job(job):
+        # job is a Future, cancel() only requests and
+        # does not currently garuntee cancellation.
+        # TODO Add a reliable cancellation mechanism
+        # (https://kutt.it/MyF1AZ)
+        return job.cancel()
+
+    @staticmethod
+    def is_active(job):
+        return not job.done()
+
+    def __init__(self, max_jobs=1):
+        self._max_jobs = max_jobs
+
+        # actually contains the Future
+        # instances from scheduled jobs
+        self._jobs = []
+
+    def refresh_jobs(self):
+        self._jobs = list(filter(
+            lambda job: JobTracker.is_active(job), self._jobs))
+
+    def __len__(self):
+        self.refresh_jobs()
+        return len(self._jobs)
+
+    def has_slots(self):
+        self.refresh_jobs()
+
+        return len(self._jobs) < self._max_jobs
+
+    def force_free_slots(self):
+        if self.has_slots():
+            return
+
+        count = 1 + len(self._jobs) - self._max_jobs
+        for job in self._jobs[:count]:
+            if not JobTracker.kill_job(job):
+                return False
+
+        return True
+
+    def prepare_slot(self, force=False):
+        """
+        JobTracker is not responsible to schedule,
+        monitor or manage the jobs in any manner, although
+        it can send a pre-empt signal to running jobs.
+        It only keeps a running count and indicates if
+        resources with respect to max allocation are
+        available to accept incoming request. Hence the
+        prepare_slot() should be called before actual
+        job allocation happens to check for resources.
+        """
+        if not self.has_slots():
+            if force is True:
+                return self.force_free_slots()
+            else:
+                return False
+
+        return True
+
+    def add(self, job):
+        # TODO self._jobs should be a dict and should
+        # map from filename to job object. That way we
+        # can prevent multiple requests fighting to run
+        # on the same source file.
+        self._jobs.append(job)
+
+
+class TrackedProcessPool:
+    """
+    Abstracts the integration of ProcessPoolExec,
+    JobTracker and func_wrapper.
+    """
+
+    def __init__(self, max_jobs=1, max_workers=1):
```

Comment:
It should generally suffice having ```max_jobs == max_workers``` with a soft upper cap of ```cpu_count() on max_workers```. Since the default for max_jobs is 1, max_workers can safely be 1.

Reply:
Hmm, maybe I misunderstand something, but isn't `max_jobs` eventually going to be a user preference? In which case, having `max_workers=multiprocessing.cpu_count()` makes sense(since that should probably never be exposed to user preferences)?

It's fine if you want to make those changes later(when it does indeed become a preference), I just thought it's a more sensible default.

Reply:
I agree with you suggestion. I think ```max_jobs``` will be a user preference sometime later. I'll defer this change until the time it is actually exposed.

Reply:
Also please review the parts of code I just added.